pkg/server/image_pull.go

```diff
@@ -17,14 +17,226 @@ limitations under the License.
+// Dependending on the remote namespace, this may be immutable or mutable.
+// While the name may differ from ref, it should itself be a valid ref.
+//
+// If the resolution fails, an error will be returned.
+func (c *criContainerdService) imageReferenceResolver(ctx context.Context, ref string) (resolvedImageName string, manifest imagespec.Manifest, compressedSize uint64, err error) {
+	var (
+		size int64
+		desc imagespec.Descriptor
+	)
+
+	// This first go routine in the local error group eg will resolve the image
+	// name; place that in the image store with another go routine; then dispatch
+	// a handler to fetch the object the manifest
+	resolver := docker.NewResolver()
+	eg, egctx := errgroup.WithContext(ctx)
+	eg.Go(func() error {
+		var (
+			egerr   error
+			fetcher remotes.Fetcher
+		)
+
+		resolvedImageName, desc, fetcher, egerr = resolver.Resolve(egctx, ref)
+		if egerr != nil {
+			return egerr
+		}
+
+		eg.Go(func() error {
+			return c.imageStore.Put(egctx, resolvedImageName, desc)
+		})
+
+		return containerd_images.Dispatch(
+			egctx,
+			remotes.FetchHandler(c.contentIngester, fetcher),
+			desc)
+	})
+
+	errs := make(chan error)
+	go func() {
+		defer close(errs)
+		errs <- eg.Wait()
+	}()
+
+	select {
+	case err = <-errs:
+		if err != nil {
+			return resolvedImageName, manifest, compressedSize, err
+		}
+	case <-egctx.Done():
+	}
+
+	image, err := c.imageStore.Get(ctx, resolvedImageName)
+	if err != nil {
+		return resolvedImageName, manifest, compressedSize,
+			fmt.Errorf("get failed for image:%q err: %v", resolvedImageName, err)
+	}
+	p, err := content.ReadBlob(ctx, c.contentProvider, image.Target.Digest)
+	if err != nil {
+		return resolvedImageName, manifest, compressedSize,
+			fmt.Errorf("readblob failed for digest:%q err: %v", image.Target.Digest, err)
+	}
+	err = json.Unmarshal(p, &manifest)
+	if err != nil {
+		return resolvedImageName, manifest, compressedSize,
+			fmt.Errorf("unmarshal blob to manifest failed for digest:%q %v", image.Target.Digest, err)
+	}
+	size, err = image.Size(ctx, c.contentProvider)
+	if err != nil {
+		return resolvedImageName, manifest, compressedSize,
+			fmt.Errorf("size failed for image:%q %v", image.Target.Digest, err)
+	}
+	compressedSize = uint64(size)
+	return resolvedImageName, manifest, compressedSize, nil
+}
+
+func (c *criContainerdService) pullImage(ctx context.Context, ref string) (imagespec.Descriptor, int64, error) {
+	var (
+		size              int64
+		desc              imagespec.Descriptor
+		resolvedImageName string
+	)
+
+	// This first go routine in the local error group eg will resolve the image
+	// name; place that in the image store with another go routine; then dispatch
+	// a handler for a sequence of handlers which: 1) fetch the object using a
+	// FetchHandler; and 3) recurse through any sub-layers via a ChildrenHandler
+	resolver := docker.NewResolver()
+	eg, egctx := errgroup.WithContext(ctx)
+	eg.Go(func() error {
```

Comment:
Don't need to start goroutine and use the error group.

Just
1) Resolve the image name;
2) Put the item into image store;
3) Use dispatch to pull the image.

This will also make it easier to write unit test.